File: 1140.stone-game-ii.js
/*
 * @lc app=leetcode id=1140 lang=javascript
 *
 * [1140] Stone Game II
 */

// @lc code=start
/**
 * @param {number[]} piles
 * @return {number}
 */
var stoneGameII = function (piles)
{
    // [0,i] 之间的和
    const prefixSum = new Array(piles.length);
    prefixSum[0] = piles[0];
    for (let i = 1; i < piles.length; i++)
    {
        prefixSum[i] = prefixSum[i - 1] + piles[i];
    }

    const cache = new Map();

    // 从 pileStartIndex 开始，指定 m，能获得的最多石头数量
    function helper(pileStartIndex, m)
    {
        const cacheKey = `${pileStartIndex}-${m}`;
        if (cache.has(cacheKey))
        {
            return cache.get(cacheKey);
        }

        let maxStoneCount = -Infinity;

        // 可以直接取走全部
        if (piles.length - pileStartIndex <= 2 * m)
        {
            maxStoneCount = getSum(pileStartIndex, piles.length - 1);
        }
        else
        {
            for (let i = 1; i <= 2 * m; i++)
            {
                // 这次拿走多少堆石头
                const currentTakeCount = getSum(pileStartIndex, pileStartIndex + i - 1);
                // 拿走之后还剩下多少石头
                const restStoneCount = getSum(pileStartIndex + i, piles.length - 1);
                // 对方最多在剩下石头里拿走多少
                const anotherTakeCount = helper(pileStartIndex + i, Math.max(i, m));

                maxStoneCount = Math.max(maxStoneCount,
                    currentTakeCount + restStoneCount - anotherTakeCount);
            }
        }
        cache.set(cacheKey, maxStoneCount);
        return maxStoneCount;
    }

    function getSum(start, end)
    {
        return prefixSum[end] - prefixSum[start] + piles[start];
    }

    const result = helper(0, 1);

    return result;
};
// @lc code=end

File: src/2026/1140.stone-game-ii.cpp
/*
 * @lc app=leetcode id=1140 lang=cpp
 *
 * [1140] Stone Game II
 */

#include <optional>
#include <vector>

// @lc code=start
class Solution {
 public:
  int stoneGameII(const std::vector<int>& piles) {
    std::vector<int> prefix_sum(piles.size());

    prefix_sum[0] = piles[0];
    for (int i = 1; i < piles.size(); i++) {
      prefix_sum[i] = prefix_sum[i - 1] + piles[i];
    }

    // memo[begin_index][last_opponent_pick_count]
    std::vector<std::vector<std::optional<GameResult>>> memo(
        piles.size(), std::vector<std::optional<GameResult>>(piles.size() + 1));

    const GameResult game_result = Play(piles, prefix_sum, 0, 1, memo);
    return game_result.player_stone_count;
  }

 private:
  struct GameResult {
    int player_stone_count;
    int opponent_stone_count;
  };

  static GameResult Play(
      const std::vector<int>& piles,
      const std::vector<int>& piles_prefix_sum,
      const int begin_index,
      const int last_opponent_pick_count,
      std::vector<std::vector<std::optional<GameResult>>>& memo) {
    if (begin_index >= piles.size()) {
      return {.player_stone_count = 0, .opponent_stone_count = 0};
    }

    if (memo[begin_index][last_opponent_pick_count]) {
      return memo[begin_index][last_opponent_pick_count].value();
    }

    // Player can pick 1 ~ (2 * last_opponent_pick_count) stone piles
    const int max_pick_count =
        std::min(2 * last_opponent_pick_count,
                 static_cast<int>(piles.size()) - begin_index);

    GameResult best_game_result = {.player_stone_count = -1,
                                   .opponent_stone_count = -1};

    for (int pick_count = 1; pick_count <= max_pick_count; pick_count++) {
      const int pick_end_index = begin_index + pick_count - 1;  // included
      const int pick_stone_count = piles_prefix_sum[pick_end_index] -
                                   piles_prefix_sum[begin_index] +
                                   piles[begin_index];

      const GameResult opponent_game_result =
          Play(piles, piles_prefix_sum, pick_end_index + 1,
               std::max(last_opponent_pick_count, pick_count), memo);
      if (opponent_game_result.opponent_stone_count + pick_stone_count >
          best_game_result.player_stone_count) {
        best_game_result = {
            .player_stone_count =
                opponent_game_result.opponent_stone_count + pick_stone_count,
            .opponent_stone_count = opponent_game_result.player_stone_count,
        };
      }
    }

    memo[begin_index][last_opponent_pick_count] = best_game_result;

    return best_game_result;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.stoneGameII({77, 12, 64, 35, 28, 4, 87, 21, 20});  // expect: 217
}
